### DISCUSSION TITLE
How do I integrate react-number-format into react-hook-form with the Material UI text field?

### DISCUSSION THREAD

**Discussion by REDLIGHTFILMS** [react-hook-form/react-hook-form#2200] [Q&A] [answered]
I can't think of any way to integrate it.Have you or have you seen examples of such integrations. Like on Codesandbox.io or something?

**Comment by bluebill1049** [comment]
Take a look at the example: https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r there is a mask component should be quite similar.

**Comment by oualidmaz** [comment, marked as answer]
Here is an example of **react-number-format + react-hook-form + Material UI**
https://codesandbox.io/s/late-feather-shr01

**Reply by zakpatterson** [reply]
I tried this out but couldn't make it behave in the expected way using the name field of that controller. I forked that sandbox and show it in two different ways. In the way you linked, the formatted string value is present in the name field. If I want the unformatted parsable number, I have to handle state myself: https://codesandbox.io/s/s7zj0

**Reply by good3n** [reply]
> Here is an example of **react-number-format + react-hook-form + Material UI**
> https://codesandbox.io/s/late-feather-shr01

Do you have an updated example not using the deprecated `as`?

**Reply by bluebill1049** [reply]
please use `render` instead, `as` will get deprecated.

**Reply by oualidmaz** [reply]
here is the same example with render : 
https://codesandbox.io/s/great-merkle-u6z49

**Reply by bluebill1049** [reply]
Thank you very much for those examples @oualidmaz

**Reply by zakpatterson** [reply]
How to get the unformatted value in handleSubmit though?

https://codesandbox.io/s/hungry-sound-8j0jn

**Reply by taschetto** [reply]
@zakpatterson https://codesandbox.io/s/practical-pateu-61jy9

**Comment by dpickett** [comment]
This worked for me and had it under hook form's state provider:

```javascript
<Controller
  render={({ onChange, value }) => (
    <NumberFormat
      allowNegative={false}
      decimalScale="2"
      onValueChange={(v) => onChange(v.value)}
      prefix="$"
      value={value}
    />
  )}
  name={name}
  control={control}
  fixedDecimalScale
/>
```

The key is wiring up the `onChange` and `value` render props.

**Reply by negibouze** [reply]
I had to make some changes.

```
<Controller
  render={({ field: { onChange, value } }) => (
    <NumberFormat
      allowNegative={false}
      decimalScale={2}
      fixedDecimalScale
      onValueChange={(v) => onChange(v.value)}
      prefix="$"
      value={value}
    />
  )}
  name={name}
  control={control}
/>
```

**Reply by wachidmudi** [reply]
Is this for v7? Cause it's not working on v6

**Reply by arpeiks** [reply]
This controller works for v7

```
<Controller
        render={({ field }) => {
          return (
            <NumberFormat
              size="small"
              color="secondary"
              variant="outlined"
              decimalScale={2}
              allowEmptyFormatting={false}
              control={control}
              prefix={"$"}
              fixedDecimalScale={true}
              thousandSeparator={true}
              {...field}
              customInput={TextField}
              onValueChange={(c) => {
                field.onChange(c.value);
              }}
            />
          );
        }}
        name="amount"
        variant="outlined"
        defaultValue="20000"
        control={control}
      />
```

https://codesandbox.io/s/heuristic-feather-03csj?file=/src/App.js

**Reply by tinhtinh95** [reply]
@arpeiks How can I reset the NumberFormat field using reset function :pray: 
I'm using inputRef={ref} but it doesn't work

**Reply by husk-9** [reply]
I tried this, but when I submit or watch the form values, the number is a string ("1.01", not `1.00`). Do you know how to fix this?

If I do something like 

```ts
onValueChange={(c) => {
	field.onChange(parseFloat(c.value));
}}
```
It messes up the react-number-format field (especially when you type zero)

**Reply by batamire** [reply]
@memeries use `c.floatValue`

**Reply by drewk** [reply]
I ran into some subtle issues with the above referenced implementations but the solution from @negibouze was very close to what we needed. The issues with that solution were:
1) I needed to pass the other props from `field` into our `NumberFormat` component, namely `name`, `onBlur`, and `ref`
2) I needed to update `onChange` to pass in the `floatValue` attribute of the `onValueChange` parameter, as mentioned by @batamire above. 
```
<Controller
  name={name}
  control={control}
  render={({ field: { onChange, onBlur, name, value, ref } }) => (
    <NumberFormat
      allowNegative={false}
      decimalScale={2}
      fixedDecimalScale
      onValueChange={(values) => onChange(values.floatValue)}
      prefix="$"
      name={name}
      value={value}
      onBlur={onBlur}
      ref={ref}
    />
  )}
/>

**Reply by alexandrzavalii** [reply]
is there a way to use register api instead?
I assume it would be more performant and avoid unnecessary updates

**Reply by ogradye** [reply]
This doesn't seem to work for me atleas unless you remove {...field} from the rendered component like 
```
<Controller
name="auction.starting_bid"
control={control}
rules={{ required: true }}
render={({ field }) => <NumberFormat
thousandSeparator={true}
prefix={'$'}
displayType="input"
type="text"
onValueChange={(values) => {
    field.onChange(values.value);
}}
className={`form-control ${errors.auction?.starting_bid ? 'is-invalid' : ''}`}
/>}
/>
``` any idea why?

**Reply by drewk** [reply]
In the above example you're using only the `onChange` prop from the render prop's `field` parameter which means you're not passing `name`, `value`, `ref`, or `onBlur` down to the `NumberFormat` component. I'm not certain how this will ultimately affect react hook form's ability to monitor your field, but I would expect things like the `touched` and `dirty` states to potentially malfunction since I believe those depend on the `onBlur` method supplied by react hook form. I do not think you want to simply spread the `field` parameter into your `NumberFormat` but rather follow something like the pattern I shared a few posts above, where I destructure the `field` parameter and pass the individual props where needed. In other words something like this:

```
<Controller
  name="auction.starting_bid"
  control={control}
  rules={{ required: true }}
  render={({ field: { onChange, onBlur, name, value, ref } }) => (
    <NumberFormat
      thousandSeparator={true}
      prefix="$"
      displayType="input"
      type="text"
      onValueChange={(values) => onChange(values.floatValue)}
      className={`form-control ${errors.auction?.starting_bid ? 'is-invalid' : ''}`}
      name={name}
      value={value}
      onBlur={onBlur}
      ref={ref}
    />
  )}
/>```

**Reply by jekh** [reply]
The answer from @drewk also seems to work when using `customInput={TextField}` from mui, but instead of passing `ref={ref}`, you need to pass `inputRef={ref}`, as mentioned here: https://github.com/s-yadav/react-number-format#getting-reference.

**Reply by Mrjuanblack** [reply]
@tinhtinh95 Were you able to reset your form? I ran into the same problem:
Controller:
```typescript
<Controller
    disabled={isLoading}
    name="price"
    control={control}
    render={({field}) =>
        <TextField
            {...field}
            error={!!errors.price}
            helperText={errors.price?.message}
            label="Price"
            margin="none"
            variant="outlined"
            fullWidth
            InputProps={{
                inputComponent: CurrencyFormatCustom as any
            }}
        />
    }
/>
```
CurrencyFormatCustom:
```typescript
interface CustomProps {
    onChange: (event: { target: { name: string; value: string } }) => void;
    name: string;
}
const CurrencyFormatCustom = React.forwardRef<NumericFormatProps, CustomProps>(
    function NumericFormatCustom(props, ref) {
      const { onChange, ...other } = props;
  
      return (
        <NumericFormat
          {...other}
          getInputRef={ref}
          onValueChange={(values) => {
            onChange({
              target: {
                name: props.name,
                value: values.value,
              },
            });
          }}
          thousandSeparator
          valueIsNumericString
          prefix="$ "
        />
      );
    },
);
```

**Reply by oussamaAbdelwahed** [reply]
> In the above example you're using only the `onChange` prop from the render prop's `field` parameter which means you're not passing `name`, `value`, `ref`, or `onBlur` down to the `NumberFormat` component. I'm not certain how this will ultimately affect react hook form's ability to monitor your field, but I would expect things like the `touched` and `dirty` states to potentially malfunction since I believe those depend on the `onBlur` method supplied by react hook form. I do not think you want to simply spread the `field` parameter into your `NumberFormat` but rather follow something like the pattern I shared a few posts above, where I destructure the `field` parameter and pass the individual props where needed. In other words something like this:
> 
> ```
> <Controller
>   name="auction.starting_bid"
>   control={control}
>   rules={{ required: true }}
>   render={({ field: { onChange, onBlur, name, value, ref } }) => (
>     <NumberFormat
>       thousandSeparator={true}
>       prefix="$"
>       displayType="input"
>       type="text"
>       onValueChange={(values) => onChange(values.floatValue)}
>       className={`form-control ${errors.auction?.starting_bid ? 'is-invalid' : ''}`}
>       name={name}
>       value={value}
>       onBlur={onBlur}
>       ref={ref}
>     />
>   )}
> />```
> ```

Hi , i'am facing an issue with isDirty , it's not being set to true when I change input, i'am instead returning a TextField from the controller render prop and setting the TextField prop inputComponent like this            ```
 inputComponent: useNumericFormatter
                ? (CustomNumericFormat as any)
                : undefined,  
``` , can you please help with that ? 
My code : 
type CustomNumericFormatProps = {
  onChange: (event: { target: { name: string; value: string } }) => void;
  onBlur: () => void;
  name: string;
};

const CustomNumericFormat = forwardRef<
  NumericFormatProps,
  CustomNumericFormatProps
>(function NumericFormatCustom(props, ref) {
  const { onChange, ...other } = props;

  return (
    <NumericFormat
      {...other}
      getInputRef={ref}
      onValueChange={(values) => {
        onChange({
          target: {
            name: props.name,
            value: values.value,
          },
        });
      }}
      valueIsNumericString
    />
  );
});

type Props = {
  control: Control<any, any>;
  name: string;
  searchBar?: boolean;
  complementaryOnChangeCallback?: (value: string) => void;
  useNumericFormatter?: boolean;
  valueAsNumber?: boolean;
  sx?: SxProps<Theme>;
} & TextFieldProps;

// eslint-disable-next-line import/prefer-default-export
export const CustomTextField: React.FC<Props> = ({
  name,
  control,
  complementaryOnChangeCallback,
  sx,
  searchBar,
  useNumericFormatter,
  valueAsNumber,
  ...rest
}) => {
  const { t, i18n }: { t: any; i18n: Ti18n } = useTranslation<'common'>();
  const locale = i18n.language;

  return (
    <Controller
      name={name}
      control={control}
      render={({
        field: { ref, onChange, ...field },
        fieldState: { error, isTouched },
        formState: { isSubmitting },
      }) => (
        <>
          <TextField
            autoComplete="off"
            sx={{ width: '100%', ...sx }}
            onChange={(event) => {
              onChange(
                valueAsNumber ? Number(event.target.value) : event.target.value
              );
              if (isFunction(complementaryOnChangeCallback))
                complementaryOnChangeCallback(event.target.value);
            }}
            {...rest}
            onInput={onChange}
            {...field}
            inputRef={ref}
            InputProps={{
              inputProps: {
                style: {
                  minHeight: '10px',
                },
                ...(useNumericFormatter
                  ? // @NB: NumericFormat accepts props named `thousandSeparator` and `decimalSeparator` (returned by getLocaleNumberSeparators)
                    getLocaleNumberSeparators(locale)
                  : {}),
              },
              startAdornment: searchBar ? (
                <InputAdornment position="start">
                  <SearchIcon color="primary" />
                </InputAdornment>
              ) : undefined,
              inputComponent: useNumericFormatter
                ? (CustomNumericFormat as any)
                : undefined,
              ...defaultTo(rest?.InputProps, {}),
            }}
            type={searchBar ? 'search' : undefined}
            inputMode={searchBar ? 'search' : undefined}
          />
          {!isSubmitting && isTouched ? (
            <Typography color="error">{t(error?.message)}</Typography>
          ) : null}
        </>
      )}
    />
  );
};

**Reply by arnoldoma** [reply]
> En el ejemplo anterior, solo se usa la `onChange`propiedad del parámetro de render, `field`lo que significa que no se pasan `name`, `value`, `ref`, ni `onBlur`al `NumberFormat`componente. No estoy seguro de cómo afectará esto a la capacidad del formulario de gancho de React para supervisar el campo, pero preveo que elementos como los `touched`estados `dirty`y podrían fallar, ya que creo que dependen del `onBlur`método proporcionado por el formulario de gancho de React. No creo que se desee simplemente distribuir el `field`parámetro en el formulario, `NumberFormat`sino seguir el patrón que compartí en algunas publicaciones anteriores, donde desestructura el `field`parámetro y pasa las propiedades individuales donde sea necesario. En otras palabras, algo como esto:
> 
> ```
> <Controller
>   name="auction.starting_bid"
>   control={control}
>   rules={{ required: true }}
>   render={({ field: { onChange, onBlur, name, value, ref } }) => (
>     <NumberFormat
>       thousandSeparator={true}
>       prefix="$"
>       displayType="input"
>       type="text"
>       onValueChange={(values) => onChange(values.floatValue)}
>       className={`form-control ${errors.auction?.starting_bid ? 'is-invalid' : ''}`}
>       name={name}
>       value={value}
>       onBlur={onBlur}
>       ref={ref}
>     />
>   )}
> />```
> ```



> En el ejemplo anterior, solo se usa la `onChange`propiedad del parámetro de render, `field`lo que significa que no se pasan `name`, `value`, `ref`, ni `onBlur`al `NumberFormat`componente. No estoy seguro de cómo afectará esto a la capacidad del formulario de gancho de React para supervisar el campo, pero preveo que elementos como los `touched`estados `dirty`y podrían fallar, ya que creo que depende del `onBlur`método proporcionado por el formulario de gancho de React. No creo que se desee simplemente distribuir el `field`parámetro en el formulario, `NumberFormat`sino seguir el patrón que compartí en algunas publicaciones anteriores, donde desestructura el `field`parámetro y pasa las propiedades individuales cuando sea necesario. En otras palabras, algo como esto:
> 
> ```
> <Controller
>   name="auction.starting_bid"
>   control={control}
>   rules={{ required: true }}
>   render={({ field: { onChange, onBlur, name, value, ref } }) => (
>     <NumberFormat
>       thousandSeparator={true}
>       prefix="$"
>       displayType="input"
>       type="text"
>       onValueChange={(values) => onChange(values.floatValue)}
>       className={`form-control ${errors.auction?.starting_bid ? 'is-invalid' : ''}`}
>       name={name}
>       value={value}
>       onBlur={onBlur}
>       ref={ref}
>     />
>   )}
> />```
> ```

**Reply by arnoldoma** [reply]
This helps me a lot. I share the solution for my integration using Controller. Maybe some of it will be useful to you.

// CONTROLLER

```
export const FieldCurrency = ({ control, name, rules, errors, ...rest }) => {
    return <Controller
        name={name}
        control={control}
        rules={rules}
        render={({ field: { label, name, decimalScale = 2, defaultValue = 0, prefix = 'Q ', onChange, onBlur, value, ref, ...field } }) => (
            <NumericFormat
                {...field}
                sx={{ mb: 1 }}
                fullWidth
                size='small'
                margin="none"
                placeholder={`${prefix} 0.00`}
                slotProps={{
                    inputLabel: {
                        shrink: true,
                    },
                }}
                customInput={TextField}
                thousandSeparator={","}
                fixedDecimalScale
                prefix={prefix}
                decimalScale={decimalScale}
                label={label}
                name={name}
                getInputRef={ref}
                value={value}
                onBlur={onBlur}
                onValueChange={(target) => {
                    onChange(target.floatValue);
                }}
                error={!!errors[name]}
                helperText={errors[name]?.message}
                {...rest}
            />
        )}
    />
}
```

// Component

```   
<FieldCurrency
      name='price'
      label='price'
      rules={{ required: 'Is required' }}
      control={control}
      errors={errors}
    />
  ```

**Comment by hasan-ceo** [comment]
Hi @bluebill1049 
My Yup code:
```
const schema = yup.object({
  amount: yup
    .number()
    .min(0, "Must be greater than or equal to 0")
    .typeError("Positive number required")
}); 
```
My controller code:
```
<Controller
        render={({ field }) => (
          <NumericFormat
            thousandSeparator
            {...field}
          />
        )}
        name="amount"
        control={control}
      />
```
Getting error: Positive number required
![image](https://user-images.githubusercontent.com/5609828/210338636-64a7b500-1967-42c0-9821-edd969f26d85.png)

**Reply by drewk** [reply]
You're probably getting a type error because of the comma in the field value. I think you might need to add a transform to your yup schema to format the input value to a valid number. I haven't done this myself but here is the relevant doc: https://github.com/jquense/yup#schematransformcurrentvalue-any-originalvalue-any--any-schema
Hope that helps.

**Reply by vaidikChouhan** [reply]
.transform((value, originalValue) =>
 parseFloat(String(originalValue || 0).replace(/,/g, ''));
      ) paste this in your schema

**Comment by bpeter340** [comment]
I can get the input field state, but validation only works for empty fields despite character length checks. What could be the issue here? @bluebill1049 

Yup Schema
````
  const schema = Yup.object().shape({
    phone_number: Yup.string()
      .required('Phone Number is required')
      .length(12, 'Must be a valid phone number (10 digits)'),
  })
````

Controller
````
<Controller
    name="phone_number"
    control={control}
    render={({ field, fieldState }) => (
      <PatternFormat
        {...field}
        label="Phone Number"
        customInput={TextField}
        type="tel"
        format="###-###-####"
        autoComplete="off"
        fullWidth
        error={!!fieldState.error?.message}
        helperText={fieldState.error?.message}
      />
    )}
  />
````

**Reply by leapful** [reply]
@bpeter340 

Please open a new discussion with codesandbox url to re-produce the issue since this current discussion is in long time ago and might be not valid with latest version of react-hook-form.
